File: migration/pset6/readability/readability.py
#!/usr/bin/env python
# Pset 6 readability.py
from cs50 import get_string


def main():
    # Prompts the user for some text
    text = input("Text: ")

    # Computes number of letters
    letters = letter_ctr(text)

    # Computes number of words
    words = word_ctr(text)

    # Computes number of sentences
    sentences = sentence_ctr(text)

    # Computes coleman-liau index
    grade_level = coleman_liau_index(letters, words, sentences)

    if grade_level < 1:
        print("Before Grade 1")
    elif grade_level >= 16:
        print("Grade 16+")
    else:
        print("Grade " + str(grade_level))


# letter_ctr()
def letter_ctr(txt):
    """Computes the number of letters in the text"""
    ctr = 0

    for i in txt:
        if i.isalpha():
            ctr += 1

    return ctr


# word_ctr()
def word_ctr(txt):
    """Extracts words from a string of text using
    the split() function"""
    words = txt.split()

    return len(words)


# sentence_ctr()
def sentence_ctr(txt):
    """Extracts the number of sentences in a
    string of text"""
    ctr = 0

    for word in txt:
        if "." in word or "!" in word or "?" in word:
            ctr += 1

    return ctr


# Computes Coleman-Liau index
def coleman_liau_index(ltrs, wrds, sen):
    """Computes the grade reading level"""
    L = (ltrs / wrds) * 100
    S = (sen / wrds) * 100

    index = .0588 * L - .296 * S - 15.8

    return int(round(index))


if __name__ == "__main__":
    main()

File: CS50/pset2/readability/readability.c
/**
 * pset 2
 * readability.c
 * February 4 2020
 * This program parses are string of text
 * and uses the Coleman-Liau index to compute
 * what grade level that the text is suited for
 */
#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <ctype.h>
#include <math.h>

// Function prototypes
float count_letters(string s);
float count_words(string s);
float count_sentences(string s);
void grade_level(string s);

int main(void)
{
    // Prompts the user to enter text
    string text = get_string("Text: ");

    // // Computes number of letters within text
    // printf("%i\n", count_letters(text));

    // // Computes the number of words within the text
    // printf("%i\n", count_words(text));

    // // Computes the number of sentences within the text
    // printf("%i\n", count_sentences(text));

    // // Uses the Coleman-Liau index to compute the grade level of the text
    grade_level(text);


    return 0;
}

/********************************************************************
 * The count_letters() function takes a string of text and iterates
 * through its characters to determine how many letters are contained
 * within it
 */
float count_letters(string s)
{
    float n = 0;

    for (int i = 0, j = strlen(s); i < j; i++)
    {
        if (isalpha(s[i]))
        {
            n++;
        }
    }

    return n;
}

/********************************************************************
 * The count_words() function takes a string of text and iterates through
 * its characters to determine how may words are contained within it
 */
float count_words(string s)
{
    float w = 0;
    int i = 0;

    while (s[i] != '\0')
    {
        if (s[i] == ' ' || s[i + 1] == '\0')
        {
            w++;
        }

        i++;
    }

    return w;
}

/********************************************************************
 * The count_sentences() function iterates through a string of text
 * to determine how many sentences are contained therein.
 */
float count_sentences(string s)
{
    float t = 0;

    for (int i = 0, n = strlen(s); i < n; i++)
    {
        if (s[i] == '.' || s[i] == '!' || s[i] == '?')
        {
            t++;
        }
    }

    return t;
}

/********************************************************************
 * The coleman-liau index computes the grade level of the text
 */
void grade_level(string s)
{
    // Gets number of words
    float words = count_words(s);

    // Gets average number of letters per 100 words
    float L = (count_letters(s) / words) * 100;

    // Gets number of sentences
    float S = (count_sentences(s) / words) * 100;

    // Computes index
    float I = 0.0588 * L - 0.296 * S - 15.8;

    if (I < 1)
    {
        printf("Before Grade 1\n");
    }

    else if (I > 16)
    {
        printf("Grade 16+\n");
    }

    else
    {
        printf("Grade %.0f\n", round(I));
    }
}